UCLALibrary/terra
# SYS-1458: Upgrade Django and move to PostgreSQL

Implements [SYS-1458 (upgrade Django)](https://uclalibrary.atlassian.net/browse/SYS-1458) and [SYS-1464 (move database from MySQL to PostgreSQL)](https://uclalibrary.atlassian.net/browse/SYS-1464).  Not tagged for release, which will happen separately when the upgrade is scheduled.

This PR implements three major changes:
1. Upgrade Django from version 2.2 to 4.2.7
2. Upgrade Python from 3.7 to 3.11
3. Change the database from MySQL to PostgreSQL

These are combined in one PR because they are closely related.  Django 4.2 no longer supports MySQL 5.7, and we're not prepared to upgrade to MySQL 8.  And switching database engines while working - even temporarily -  with the long-unsupported Django 2.2  / Python 3.7 is highly undesirable.

Despite these major changes, the code itself did not require much updating.  Highlights:
* The Docker-related files were updated to match our current practices.
* Relevant values in `settings.py` were added / updated as needed.
* Libraries in `requirements.txt` were updated, and locked to specific current versions where relevant.
  * In particular, `crispy-forms` was updated from 1.x to 2.1, and the related `bootstrap` template package updated from 4.0 to 5.0.  I reviewed all output and confirmed there are no (noticeable) visual changes.
  * Developer-specific libraries (`black`, `pre-commit`, and `pyflakes`) do not have version specified.  These are a remnant of old development practices and probably will be removed later.
*  A MySQL-specific custom override of `test.py` was removed; we now use the default Django test runner.
* A change in Django 3.2 required updating `Value(0)` (which is now treated as an `int`) to `Decimal(0)`, for compatibility with Django's `DecimalField` type.
* Minor changes to `admin.py` due to Django changing how short descriptions are specified.
* Added our standard `favicon` to the base template.

### Testing
I mainly want to be sure this application builds and runs for more than just me 😟 .  Please do the following:
* Clone the repo and checkout `SYS-1458/upgrade_django_terra`
* `docker-compose up --build`
* Don't worry about the lengthy output from the `Dockerfile` build process, but check the `django` and `db` output for errors.
* In a separate terminal window, run the tests - should be 81, all passing: `docker-compose exec django python manage.py test`
* In a browser, go to http://127.0.0.1:8000/
  * Log in as one of our test users: `doriswang`, password `Staples50141`
  * Click around the various options in the top navigation bar.
  * Don't worry about the data, which is minimal and old.  Just be sure everything looks consistent, with no error messages.

Thanks --Andy
